ENH: alternative (ctypes) approach to openblas version access

tylerjereddy opened the pull request:
This is the alternative approach to #12523, which currently has a lot of debug code and actually breaks the Windows builds --  you can take a look in the Azure output if you want.

I've not had success getting this to work on Windows, even if I put all the DLL / lib files at the same path.

Open to suggestions, otherwise the extension building in the original PR may be preferred so that we can be agnostic to static / dynamic details. Not sure any approach is really going to be fully portable to all the scenarios that other libraries may have for openblas version access.

tylerjereddy commented:
credit to @mattip for suggesting this approach

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -20,14 +22,17 @@ jobs:
+           CFLAGS='-UNDEBUG -std=c99' python3 runtests.py --mode=full -- -rsx --junitxml=junit/test-results.xml && \
+           cd .. && \
+           export PYTHONPATH=/numpy/build/testenv/lib/python3.6/site-packages && \
+           python3 -c \"import numpy; from numpy.distutils.system_info import get_info; assert get_info('openblas')['versions'][0] == '$(OPENBLAS_VERSION)', 'received versions: ' + ' '.join(get_info('openblas')['versions']) + ' and expected $(OPENBLAS_VERSION)'\""
```
this "just works" for linux, which is great I think

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -100,6 +105,7 @@ jobs:
       # appveyor / Windows config
       OPENBLAS_32: "https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-v0.3.3-186-g701ea883-win32-gcc_7_1_0.zip"
       OPENBLAS_64: "https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-v0.3.3-186-g701ea883-win_amd64-gcc_7_1_0.zip"
+      OPENBLAS_VERSION: "0.3.6.dev"
```
intentionally using the wrong value here for now -- if we do get Windows to work I want to see the version mismatch detection first before correcting

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -151,8 +157,10 @@ jobs:
+      $target_dll = "C:\\hostedtoolcache\\windows\\Python\\$pyversion\\$(PYTHON_ARCH)\\lib\\openblas.dll"
       Write-Host "target path: $target"
       cp $tmpdir\$(BITS)\lib\libopenblas_v0.3.3-186-g701ea883-gcc_7_1_0.a $target
+      cp $tmpdir\$(BITS)\bin\libopenblas_v0.3.3-186-g701ea883-gcc_7_1_0.dll $target_dll
```
this **really** needs to be clarified; while working on both NumPy / SciPy CIs I've noticed that we mostly just tend to move the static library file and build using that; so, I'm not clear on what exactly is going on with dynamic openblas builds & if we can viably use the DLL for version detection; just copying the DLL file in causes the Windows builds to break

mattip commented:
How do I test this locally? Which powershell script with what local environment variables should I use so `runtests` finds the openblas you wish to test?

mattip commented:
Understood. But in order to run the test it needs to find some blas library, any one. How do I get started from scratch so I can run the new test and see what is wrong, why windows fails to load the dll?

tylerjereddy commented:
There's already traceback info in the Azure output -- I would just debug in a fork using Azure runs, which are basically instant.

If you really want local for fastest iteration times, I'd make sure to start with a "naive" Windows configuration -- i.e., `depends.exe` can't work with the downloaded MacPython Windows wheel bin/openblas.dll file, and start putting those MacPython lib files in the lib path used to build NumPy.

mattip commented:
The dll in the [package](https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-v0.3.3-186-g701ea883-win_amd64-gcc_7_1_0.zip) downloaded by the CI  requires `LIBGCC_S_SEH-1.DLL` and `LIBGFORTRAN-4.DLL`. So it seems we cannot report the OpenBLAS version during initial OpenBLAS configuation, only after building.

tylerjereddy commented:
The requirement for post-build version reporting is also a constraint on the static approach, although POSIX DLL approach can grab the version right away, which is quite nice.

How can I extract those missing DLL requirements programatically using Python? As far as I can tell, retrieving the DLL dependencies of DLLs in a robust manner is tricky at best, even without the constraint of a Python hook.

Is there any reason I can't put those DLLs on an appropriate library search path before the build? The fragility does concern me I guess.

mattip commented:
> How can I extract those missing DLL requirements programatically using Python

You need to find some source for them. I think they are part of mingw's gfortran

tylerjereddy commented:
Right, but we can't actually naively probe a DLL on Windows with Python & have it tells us the names of the files that are missing? `ctypes` tends to produce a generic error as does the depends.exe.

tylerjereddy commented:
Some updates on the Azure end now allow detection of the `OpenBLAS` version dynamically in the Windows 64-bit builds. Probably an update to the mingw gcc toolchain and / or path adjustments to find the appropriate DLLs automatically.

We recently pinned 32-bit Windows builds to an older mingw / gcc toolchain for other reasons, but can maybe just manually patch the path issues if that's the problem for DLL resolution.

I have to debug on my fork / another branch first though.

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -164,13 +207,19 @@ jobs:
+      $pyversion = python -c "from __future__ import print_function; import sys; print(sys.version.split()[0])"
+      $target_include = Resolve-Path "C:\\hostedtoolcache\\windows\\Python\\$pyversion\\$(PYTHON_ARCH)\\include\\" | Select -ExpandProperty Path
+      $env:INCLUDE = $target_include + ";" + $env:INCLUDE 
+      echo $env:INCLUDE
```
this doesn't seem to help with the symbol resolution, but was worth a try I guess

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -149,9 +155,46 @@ jobs:
+         lib /def:openblas.def /out:openblas.lib /machine:x86
+      }
+      If ($(BITS) -eq 64) {
+         lib /def:openblas.def /out:openblas.lib /machine:x64
```
This should probably happen in MacPython wheel builds, I think, if it is truly needed. Certainly the lack of a `lib` file was holding back the builds before I did this.

tylerjereddy commented:
@mattip If you check the compile / link commands just before `zungqr_` is missing in the Windows NumPy builds above, perhaps you can give me some suggestions.

Clearly that object / symbol is all over the place in the openblas files, so hopefully just something simple is missing. Maybe we really need to generate that `openblas.lib` properly instead of in a hacky way.

tylerjereddy committed:
ENH: add compile-time OpenBLAS version detection

* add a ctypes-based approach for compile-time
detection of the version of OpenBLAS that is linked
with NumPy

* on Windows, this will require usage of openblas.dll,
the import library (.dll.a renamed to .lib), and the
usual static library (for now, at least)

* added unit tests; also added checks in Azure CI for
Windows, Linux, and MacOS jobs to ensure that we are using
appropriate version of OpenBLAS

tylerjereddy commented:
Compile-time OpenBLAS version detection should now be working in this PR based on my fork-based Azure runs. That means OpenBLAS version is confirmed for all NumPy builds on Azure Windows and Azure Linux in this PR.

I've left a few verbose comments in the code for now. This took a substantial amount of debugging!

We are currently chained to a workflow that requires 3 OpenBLAS-related library files to get both successful build / test **and** compile-time OpenBLAS version detection on **Windows**:
- the DLL
- the (stub) import library for the DLL
- the static library proper, which seems to be fundamentally needed  by distutils machinery to generate i.e., `numpy/.libs` DLL, etc.

I'm not sure how hard it would be (or if it is even desirable) to completely abrogate the need for linking to the static `openblas.a`.

Potential concerns:
- this will now generate two DLLs in `numpy/.libs` (there's no evidence this causes problems -- and Azure CI will now have three DLLs there sometimes because of the need for handling this behavior previously with conda issue in last few days)
- even more disruption to Windows DLL handling in NumPy after recent issues, but if we're afraid to make improvements that's a problem too

tylerjereddy commented:
To see it in action at compile (wheel build) time, look for entries like this in the Azure Build NumPy stages:

```
    FOUND:
      library_dirs = ['C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\pip-req-build-80xrtfay\\build\\openblas', 'c:\\hostedtoolcache\\windows\\python\\3.6.8\\x86\\lib', 'C:\\', 'c:\\hostedtoolcache\\windows\\python\\3.6.8\\x86\\libs']
      libraries = ['openblas', 'openblas']
      language = f77
      define_macros = [('HAVE_CBLAS', None)]
      versions = ['0.3.5.dev']
```

tylerjereddy commented:
I've removed the WIP tag since the OpenBLAS versions are showing up on Linux / Windows and CI is all green--anything I can do to alleviate concerns/ move this forward?

tylerjereddy commented:
This needs to be updated after the merging of #12457 -- I might as well demonstrate that we can get compile-time OpenBLAS version on MacOS as well.

tylerjereddy commented:
Updated to include OpenBLAS compile-time & post-mortem version detection on MacOS in Azure, so all of the Azure matrix entries should demonstrate this now.

This also revealed one weakness in the MacOS test workflow--pip install of matplotlib (for refguide) has to be moved after the dev build to avoid a double install of NumPy (1.16.2 alongside 1.17-dev). This may have been relatively harmless before, but not anymore, so I've made that change too.

tylerjereddy committed:
MAINT: try bumping OpenBLAS to v0.3.5

tylerjereddy commented:
The most recent commit aims to bump from OpenBLAS `0.3.5.dev` to `0.3.5`, to reflect the recent [MacPython availability of the latest stable release](https://github.com/MacPython/openblas-libs/pull/1).

Specifically, the objective is to ensure the version detection machinery appears robust across an actual version change in the wild.

tylerjereddy commented:
@stefanv raised some concerns about "hot-swapping" the linalg backend as it pertains to the infrastructure change here. Specifically, building NumPy with one linear algebra backend, and then changing to another at runtime.

 I'm not clear on the failure scenario, or how this PR would make it any better / worse.

This provides a way to probe whatever OpenBLAS is present, both at compile and run time, whether you install from a wheel or not. From the perspective of the "dynamic" OpenBLAS library, it is always runtime, but from NumPy's perspective it could be either compile or runtime--you call the same C function in pre-built OpenBLAS.

I can't really comment on the advisability of building NumPy locally with say OpenBLAS backend, then removing access to those libs & adding another backend at runtime, or even having no backend and hoping NumPy falls back to lapack_lite gracefully at runtime after having been linked with say OpenBLAS.

For wheels, it so happens that after a user installs the wheel, the OpenBLAS version that they probe with the functionality provided should be a true match to the version the wheel was built with, because of the way dependencies like OpenBLAS runtime libs are included with the wheels.

For end users reporting issues for source builds, if there really are people in environments where they build with one lib and maybe even accidentally use another at runtime, we can still ask for the appropriate logs--since the machinery can be used at compile (directly printed by distutils) and runtime (from `system_info`)--a difference might signal an issue.

Am I missing something here?

rgommers commented:
> @stefanv raised some concerns about "hot-swapping" the linalg backend as it pertains to the infrastructure change here. Specifically, building NumPy with one linear algebra backend, and then changing to another at runtime.

I wouldn't call this hot-swapping (that would be after importing numpy), but yes this is possible in principle. Conda-forge just made some changes that enable exactly this; install numpy, and then switch between BLAS libs: https://github.com/conda-forge/numpy-feedstock/pull/143

mattip commented:
Should we close this?

tylerjereddy commented:
I'll close it because of the lack of interest & because people seem to like dropping open PR numbers.

But it still highlights a useful weakness--we can't probe Windows OpenBLAS versions via ctypes without intervening in distutils, probably because of the assumption of the usage of mingw virtual static library openblas.a and special handling around that.
